In [37]:
import pandas as pd

In [38]:
#Reading in Data 

#creating dataframe of metadata of recipes called recipes_df
recipes_df = pd.read_csv("./Data/RAW_recipes.csv")

#creating dataframe of user reviews called user_review_df
user_reviews_df = pd.read_csv("./Data/RAW_interactions.csv")

In [39]:
recipes_df.head()

name      id  minutes  \
0  arriba   baked winter squash mexican style  137739       55   
1            a bit different  breakfast pizza   31490       30   
2                   all in the kitchen  chili  112140      130   
3                          alouette  potatoes   59389       45   
4          amish  tomato ketchup  for canning   44061      190   

   contributor_id   submitted  \
0           47892  2005-09-16   
1           26278  2002-06-17   
2          196586  2005-02-25   
3           68585  2003-04-14   
4           41706  2002-10-25   

                                                tags  \
0  ['60-minutes-or-less', 'time-to-make', 'course...   
1  ['30-minutes-or-less', 'time-to-make', 'course...   
2  ['time-to-make', 'course', 'preparation', 'mai...   
3  ['60-minutes-or-less', 'time-to-make', 'course...   
4  ['weeknight', 'time-to-make', 'course', 'main-...   

                                    nutrition  n_steps  \
0       [51.5, 0.0, 13.0, 0.0, 2.0, 0.0, 4.0]       11   
1   [173.4, 18.0, 0.0, 17.0, 22.0, 35.0, 1.0]        9   
2  [269.8, 22.0, 32.0, 48.0, 39.0, 27.0, 5.0]        6   
3   [368.1, 17.0, 10.0, 2.0, 14.0, 8.0, 20.0]       11   
4   [352.9, 1.0, 337.0, 23.0, 3.0, 0.0, 28.0]        5   

                                               steps  \
0  ['make a choice and proceed with recipe', 'dep...   
1  ['preheat oven to 425 degrees f', 'press dough...   
2  ['brown ground beef in large pot', 'add choppe...   
3  ['place potatoes in a large pot of lightly sal...   
4  ['mix all ingredients& boil for 2 1 / 2 hours ...   

                                         description  \
0  autumn is my favorite time of year to cook! th...   
1  this recipe calls for the crust to be prebaked...   
2  this modified version of 'mom's' chili was a h...   
3  this is a super easy, great tasting, make ahea...   
4  my dh's amish mother raised him on this recipe...   

                                         ingredients  n_ingredients  
0  ['winter squash', 'mexican seasoning', 'mixed ...              7  
1  ['prepared pizza crust', 'sausage patty', 'egg...              6  
2  ['ground beef', 'yellow onions', 'diced tomato...             13  
3  ['spreadable cheese with garlic and herbs', 'n...             11  
4  ['tomato juice', 'apple cider vinegar', 'sugar...              8

In [7]:
user_reviews_df.head()

user_id  recipe_id        date  rating  \
0    38094      40893  2003-02-17       4   
1  1293707      40893  2011-12-21       5   
2     8937      44394  2002-12-01       4   
3   126440      85009  2010-02-27       5   
4    57222      85009  2011-10-01       5   

                                              review  
0  Great with a salad. Cooked on top of stove for...  
1  So simple, so delicious! Great for chilly fall...  
2  This worked very well and is EASY.  I used not...  
3  I made the Mexican topping and took it to bunk...  
4  Made the cheddar bacon topping, adding a sprin...

In [40]:
user_reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1132367 entries, 0 to 1132366
Data columns (total 5 columns):
 #   Column     Non-Null Count    Dtype 
---  ------     --------------    ----- 
 0   user_id    1132367 non-null  int64 
 1   recipe_id  1132367 non-null  int64 
 2   date       1132367 non-null  object
 3   rating     1132367 non-null  int64 
 4   review     1132198 non-null  object
dtypes: int64(3), object(2)
memory usage: 43.2+ MB


In [43]:
user_reviews_df["recipe_id"].value_counts()

2886      1613
27208     1601
89204     1579
39087     1448
67256     1322
          ... 
150097       1
143966       1
238188       1
244335       1
4098         1
Name: recipe_id, Length: 231637, dtype: int64

In [42]:
# Check how many users 
user_reviews_df["user_id"].value_counts()

424680        7671
37449         5603
383346        4628
169430        4076
128473        3917
              ... 
2000504285       1
1934789          1
1988031          1
2067844          1
4098             1
Name: user_id, Length: 226570, dtype: int64

We have 1,132,367 user reviews on 231,637 recipes from 226,570 reviewers. 

In [ ]:
# See the spread of ratings
user_reviews_df["rating"].value_counts()

In [ ]:
#Looking at the 0 ratings to see if they are actually 0s
user_reviews_df.loc[user_reviews_df["rating"] == 0]

In [44]:
# Modeling requires that recipes have at least 5 reviews
# Checking how many reviews each recipe has

user_reviews_df["recipe_id"].value_counts()

2886      1613
27208     1601
89204     1579
39087     1448
67256     1322
          ... 
150097       1
143966       1
238188       1
244335       1
4098         1
Name: recipe_id, Length: 231637, dtype: int64

In [45]:
# Creating a list of recipe_review_counts for each recipe
# Remove recipes with less than 5 review counts

recipe_review_counts = user_reviews_df["recipe_id"].value_counts()

user_reviews_df = user_reviews_df[user_reviews_df["recipe_id"].isin(recipe_review_counts[recipe_review_counts >= 5].index)]

# Check that recipes with under 5 reviews were removed

user_reviews_df["recipe_id"].value_counts()


2886      1613
27208     1601
89204     1579
39087     1448
67256     1322
          ... 
138982       5
81292        5
77198        5
224710       5
292646       5
Name: recipe_id, Length: 53593, dtype: int64

In [46]:
# See what percentage of recipes our new dataset has compared to original
(len(user_reviews_df["recipe_id"].value_counts()))/231637

0.23136631885234224

23% of our recipes in our dataset has at least 5 reviews

In [47]:
# Checking the amount of reviews per user. 
# For recommendation modeling we need at least 5 reviews per reader. 

user_reviews_df["user_id"].value_counts()

37449         3019
383346        2570
424680        2457
128473        2175
89831         2028
              ... 
1802453203       1
1018058          1
519695           1
2001727685       1
4098             1
Name: user_id, Length: 186250, dtype: int64

In [48]:
# Create a list review counts for each user 
# Keep only those with more than 5 in the dataframe 
review_counts = user_reviews_df["user_id"].value_counts()
user_reviews_df = user_reviews_df[user_reviews_df["user_id"].isin(review_counts[review_counts >= 5].index)]

# Check that users with under 5 reviews were removed
user_reviews_df["user_id"].value_counts()

37449     3019
383346    2570
424680    2457
128473    2175
89831     2028
          ... 
609784       5
348792       5
201392       5
369766       5
623055       5
Name: user_id, Length: 18964, dtype: int64

In [49]:
# Check out the percent of our unique reviewers in our new dataframe from the original number of reviewers
(len(user_reviews_df["user_id"].value_counts()))/226570

0.08370040164187668

Our unique reviewers dropped from 226,570 reviewers to 18,964 reviewers, which means we are left with roughly 9% of unique reviewers from our original dataset.

In [50]:
user_reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 598158 entries, 8 to 1132357
Data columns (total 5 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   user_id    598158 non-null  int64 
 1   recipe_id  598158 non-null  int64 
 2   date       598158 non-null  object
 3   rating     598158 non-null  int64 
 4   review     598154 non-null  object
dtypes: int64(3), object(2)
memory usage: 27.4+ MB


In [51]:
len(user_reviews_df)/1132367

0.5282368702019752

Our new dataframe that contains recipes with 5 or more reviews and reviewers with 5 or more reviews contains 52% reviews of our original dataset. Our unique users are 9% of our original dataset, and our recipes are nearly 25% of our original dataset. While this is a slimmed down version, there is sufficient data for modeling. 

Our new dataset has four missing values for review category. 

In [53]:
# Save off the trimmed down dataset as a new csv_file 

user_reviews_df.to_csv("./Data/user_reviews.csv")

### Visualizing Trimmed Down Dataset

In [65]:
user_reviews_df["rating"].value_counts(normalize=True)

5    0.770599
4    0.158886
3    0.032045
0    0.024564
2    0.009275
1    0.004631
Name: rating, dtype: float64

77% of our new dataset includes 5 star reviews, which heavily skews the data and will affect the recommendation portal. 

We also notice that 2.4% of our dataset includes a 0 rating. It would be important to see if these are actually 0s or if they are imputed value for missing ratings. 

In [63]:
#Looking at the 0 ratings to see if they are actually 0s
user_reviews_df.loc[user_reviews_df["rating"] == 0]

user_id  recipe_id        date  rating  \
12        468945     134728  2008-02-20       0   
23        681408     225241  2008-03-14       0   
56        980899      79222  2010-10-18       0   
539        93911     306919  2010-07-08       0   
642       921361     232042  2014-10-03       0   
...          ...        ...         ...     ...   
1131752   194886      99731  2013-01-24       0   
1131780   113509     130019  2012-04-20       0   
1131953   374506      39724  2008-03-19       0   
1132016   329378      39902  2006-10-19       0   
1132031   811179     136702  2011-12-29       0   

                                                    review  
12       Made my own buttermilk w/ vinegar and milk.  U...  
23       This is a very good recipe.  We also want to c...  
56       I used half cream and half 1% milk, added anot...  
539      I haven't tried this recipe so I'm not rating ...  
642      Tasted great and was super easy to make. Thank...  
...                                                    ...  
1131752  This is a great recipe, I thought I had review...  
1131780  Hubby and I thought this was good. Neither of ...  
1131953  Great!!! After the turkey was done I browned t...  
1132016  While the recipe calls for 3 cups of cheese, t...  
1132031  We made our gluten-free with GF flour and it w...  

[14693 rows x 5 columns]

It looks like the 0s are both a mix of positive and negative (or neutral reviews). While it is a small subset of our dataset it could be important to include these reviews considering the skew of our data. We could impute ratings based on the language used in the "review" column. To do this we will need to undertake Natural Language Processing (NLP).  

In [ ]:
# Check to see which users are always high reviewers and low reviewers
#then we can decide if we need to remove